### PR TITLE
#1090: Notification to the consumer when the product is back in stock

### PR DESCRIPTION
**Pull request by s04v**
#1090 Implementation

### CODE REVIEW

File: src/Modules/SimplCommerce.Module.Catalog/Areas/Catalog/Views/Product/ProductDetail.cshtml [hunk 1]
```diff
@@ -199,6 +199,31 @@
                     <div class="out-of-stock">
                         <span class="label label-danger">@Localizer["Out of stock"]</span>
                     </div>
+
+                    <div class="back-in-stock-subscribe">
+                        <form action="api/stocks/back-in-stock">
+                            <div>
+                                @if (SignInManager.IsSignedIn(User))
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Catalog/Areas/Catalog/Views/Product/ProductDetail.cshtml:206]:
```suggestion
                                @if (User.Identity.IsAuthenticated)
```

File: src/Modules/SimplCommerce.Module.Core/Extensions/WorkContext.cs [hunk 2]
```diff
@@ -33,6 +33,11 @@ public class WorkContext : IWorkContext
             _configuration = configuration;
         }
 
+        public string GetCurrentHostName()
+        {
+            return _httpContext.Request.Host.Value;
+        }
+
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Core/Extensions/WorkContext.cs:40]:
```suggestion
        public string GetCurrentHostName() => _httpContext.Request.Host.Value;

```

File: src/Modules/SimplCommerce.Module.Inventory/Areas/Inventory/Controllers/StockApiController.cs [hunk 3]
```diff
@@ -135,5 +139,21 @@ public async Task<IActionResult> GetStockHistory(int warehouseId, int productId)
...
             return Ok(stockHistory);
         }
+
+        [AllowAnonymous]
+        [HttpPost("back-in-stock")]
+        public async Task<IActionResult> BackInStockSubscribe(long productId, string customerEmail)
+        {
+            if (await _backInStockSubscriptionRepository.Query()
+                .Where(o => o.ProductId == productId && o.CustomerEmail == customerEmail)
+                .AnyAsync())
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Areas/Inventory/Controllers/StockApiController.cs:149]:
```suggestion
            if (await _backInStockSubscriptionRepository.Query()
                .AnyAsync(o => o.ProductId == productId && o.CustomerEmail == customerEmail))
```

File: src/Modules/SimplCommerce.Module.Inventory/Event/BackInStock.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using MediatR;
+
+namespace SimplCommerce.Module.Inventory.Event
+{
+    public class BackInStock : INotification
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Event/BackInStock.cs:10]: ProductBackInStock might more meaningful

File: src/Modules/SimplCommerce.Module.Inventory/Event/BackInStockSendEmailHandler.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+namespace SimplCommerce.Module.Inventory.Event
+{
+    public class BackInStockSendEmailHandler : INotificationHandler<BackInStock>
+    {
+        public readonly IStockSubscriptionService _stockSubscriptionService;
+
+        public BackInStockSendEmailHandler(IStockSubscriptionService stockSubscriptionService)
+        {
+            _stockSubscriptionService = stockSubscriptionService;
+        }
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Event/BackInStockSendEmailHandler.cs:19]:
```suggestion
        public BackInStockSendEmailHandler(IStockSubscriptionService stockSubscriptionService)
            => _stockSubscriptionService = stockSubscriptionService;
```

File: src/Modules/SimplCommerce.Module.Inventory/Event/BackInStockSendEmailHandler.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+        public async Task Handle(BackInStock notification, CancellationToken cancellationToken)
+        {
+            await _stockSubscriptionService.BackInStockSendNotificationsAsync(notification.ProductId);
+        }
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Event/BackInStockSendEmailHandler.cs:24]:
```suggestion
        public async Task Handle(BackInStock notification, CancellationToken cancellationToken)
            => await _stockSubscriptionService.BackInStockSendNotificationsAsync(notification.ProductId);
```

File: src/Modules/SimplCommerce.Module.Inventory/Models/BackInStockSubscription.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using SimplCommerce.Infrastructure.Models;
+
+namespace SimplCommerce.Module.Inventory.Models
+{
+    public class BackInStockSubscription : EntityBase
+    {
+        public long ProductId { get; set; }
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Models/BackInStockSubscription.cs:12]:
```suggestion
        public long ProductId { get; set; }

```

File: src/Modules/SimplCommerce.Module.Inventory/ModuleInitializer.cs [hunk 8]
```diff
@@ -12,6 +16,9 @@ public class ModuleInitializer : IModuleInitializer
         public void ConfigureServices(IServiceCollection serviceCollection)
         {
             serviceCollection.AddTransient<IStockService, StockService>();
+            serviceCollection.AddTransient<IStockSubscriptionService, StockSubscriptionService>();
+            serviceCollection.AddTransient<INotificationHandler<BackInStock>, BackInStockSendEmailHandler>();
+
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/ModuleInitializer.cs:21]:
```suggestion
```

File: src/Modules/SimplCommerce.Module.Inventory/Services/StockService.cs [hunk 9]
```diff
@@ -44,6 +48,8 @@ public async Task UpdateStock(StockUpdateRequest stockUpdateRequest)
             var product = await _productRepository.Query().FirstOrDefaultAsync(x => x.Id == stockUpdateRequest.ProductId);
             var stock = await _stockRepository.Query().FirstOrDefaultAsync(x => x.ProductId == stockUpdateRequest.ProductId && x.WarehouseId == stockUpdateRequest.WarehouseId);
 
+            var prevStockQuantity = product.StockQuantity;
+
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Services/StockService.cs:52]:
```suggestion
```

File: src/Modules/SimplCommerce.Module.Inventory/Services/StockSubscriptionService.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        public async Task BackInStockSubscribeAsync(long productId, string customerEmail)
+        {
+            var subscription = new BackInStockSubscription
+            {
+                ProductId = productId,
+                CustomerEmail = customerEmail
+            };
+
+            _backInStockSubscriptionRepository.Add(subscription);
+            await _backInStockSubscriptionRepository.SaveChangesAsync();
```
Comment (hishamco) [inline src/Modules/SimplCommerce.Module.Inventory/Services/StockSubscriptionService.cs:67]:
```suggestion

            await _backInStockSubscriptionRepository.SaveChangesAsync();
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
@thiennn I suggested some improvements to this PR also, there 're some non meaningful names, but seems you ignored all the my code suggestions

**Comment by s04v** [comment]
@hishamco I can make your improvements in the new PR

**Comment by hishamco** [comment]
Believe it or not, it doesn't require a new PR, that's why the reviewing is helpful otherwise it's a waste of time if no one consider it

Don't get me wrong

**Comment by thiennn** [comment]
Hi @hishamco , I am sorry. I was on a mobile and I didn't see your suggestion.

@s04v Yes, please help me make the new PR, Thanks

**Pull request merged**